File: basic-pest/json2simpletxt.py
#!/usr/bin/env python

import json
import glob

files = sorted( glob.glob('%s/*.json' % "/tmp/dvmdostem/calibration/yearly/") )

with open(files[-1]) as f:
  fdata = json.load(f)

# PFT variables
pftvars = ['GPPAll','NPPAll','VEGCL','VEGCW','VEGCR','VEGCSUM','VEGNL','VEGNW','VEGNR','VEGNSUM','VEGLBLN']
pftvarskeys = ['GPPAll', 'NPPAll', 'VEGCarbon']

with open('basicpestsimpleoutput.csv', 'w') as f:
  f.write("Variable,Value\n")
  for i in range(0, 9):
    pftkey = 'PFT%i' % (i)
    f.write('GPPAll%s,%s\n' % (i, fdata[pftkey]['GPPAll']))
    f.write('NPPAll%s,%s\n' % (i, fdata[pftkey]['NPPAll']))
    f.write('VEGCL%s,%s\n' % (i, fdata[pftkey]['VegCarbon']['Leaf']))
    f.write('VEGCW%s,%s\n' % (i, fdata[pftkey]['VegCarbon']['Stem']))
    f.write('VEGCR%s,%s\n' % (i, fdata[pftkey]['VegCarbon']['Root']))
    f.write('VEGCSUM%s,%s\n' % (i, fdata[pftkey]['VegCarbon']['Leaf'] + fdata[pftkey]['VegCarbon']['Stem'] + fdata[pftkey]['VegCarbon']['Root']))
    f.write('VEGNL%s,%s\n' % (i, fdata[pftkey]['VegStructuralNitrogen']['Leaf']))
    f.write('VEGNW%s,%s\n' % (i, fdata[pftkey]['VegStructuralNitrogen']['Stem']))
    f.write('VEGNR%s,%s\n' % (i, fdata[pftkey]['VegStructuralNitrogen']['Root']))
    f.write('VEGNSUM%s,%s\n' % (i, fdata[pftkey]['VegCarbon']['Leaf'] + fdata[pftkey]['VegCarbon']['Stem'] + fdata[pftkey]['VegCarbon']['Root']))
    f.write('VEGLBLN%s,%s\n' % (i, fdata[pftkey]['VegLabileNitrogen']))
    

'''
GPPAll0,110.7451347285
NPPAll0,55.3725673643
VEGCL0,50.90577375
VEGCW0,87.334353575
VEGCR0,40.0896369243
VEGCSUM0,178.3297642493
VEGNL0,2.1151769691
VEGNW0,1.3782680325
VEGNR0,0.8439923563
VEGNSUM0,4.3374373579
VEGLBLN0,0.9190853995
'''



'''

{u'GPPAll': 28.1762950271368,
 u'GPPAllIgnoringNitrogen': 28.1762950271368,
 u'InNitrogenUptake': -933324.09375,
 u'LitterfallCarbonAll': 16.7496477663517,
 u'LitterfallNitrogen': {u'Leaf': -933324.09375,
  u'Root': -933324.09375,
  u'Stem': -933324.09375},
 u'LitterfallNitrogenPFT': 0.0,
 u'LuxNitrogenUptake': -933324.09375,
 u'NPPAll': 16.76845213770866,
 u'NPPAllIgnoringNitrogen': 16.76845213770866,
 u'PARAbsorb': 2.737188942392885,
 u'PARDown': 4.587406726596138,
 u'StNitrogenUptake': 0.0,
 u'TotNitrogenUptake': -933324.09375,
 u'VegCarbon': {u'Leaf': 8.41254997253418,
  u'Root': 2.343128681182861,
  u'Stem': 37.53537750244141},
 u'VegLabileNitrogen': 0.001000000047497451,
 u'VegStructuralNitrogen': {u'Leaf': -22399848.0,
  u'Root': -22399848.0,
  u'Stem': -22399848.0}
  }
'''
# for i in range(0,9):
#   pftkey = 'PFT%s' % (i)
#   for var in pftvars:
#     print fdata[pftkey][var]

from IPython import embed; embed()

File: basic-pest/json2simpletxt.r
#!/usr/bin/env Rscript

# Script to translate json dvmdostem output into a 
# csv (text) file that can be read by PEST.

# Currently set up for pft-specific pools only, for a boreal 
# shrub community with 9 pfts. To add soil pools, would need to 
# add code for extracting appropriate variables from the json 
# object then adding appropriate variable names 
# and values to the obs.vals data.frame produced at the end.

# Results are written out into a .txt file comma delimited.
# output can be found in the outdir

library(rjson)
#config <- fromJSON(file="../dvm-dos-tem/config/config.js")

#print(config)

# SET THE CASE NAME HERE:
casename<-"basicpest"

# store the current working directory so we can write the output file here
script_run_from_directory<-getwd()
ofname<-file.path(script_run_from_directory, paste(casename, "simpleoutput.csv", sep=""))

# do the parsing in the the /tmp directory...
setwd('/tmp/dvmdostem/calibration/yearly')

# haven't implemented this yet- want to take the averages of last 10 years model output for comparison.
avg.years=10
jsons=dir(pattern='json')
num.files=length(jsons)
print("Found this many json files:")
print(num.files)
last.jsons=jsons[(num.files-avg.years):num.files]
######################################################

cmtbgc<-fromJSON(file=jsons[num.files]) #extracting last year json file as object 'cmtbgc'

#pftspecific code- generating a matrix with parameters as rows and pfts as columns
# there is also code to label pft-specific target variables with the appropriate pft index
pft.nums<-0:8 #this can be changed to any vector of pft numbers depending on the community

pft.ids<-paste('PFT',pft.nums,sep='')
num.pfts=length(pft.ids)

#the following vector should be changed depending on which veg-specific target variables you are using.
pft.vars<-c('GPPAll','NPPAll','VEGCL','VEGCW','VEGCR','VEGCSUM','VEGNL','VEGNW','VEGNR','VEGNSUM','VEGLBLN')

#this loop extracts pft-specfic output variables from cmtbgc json object.
pft.vals.mat<-matrix(0,length(pft.vars),num.pfts) #values go into this matrix.
for(j in 1:num.pfts) 
{
  pft=pft.ids[j] #getting the pft index number
  pft.dat<-cmtbgc[[pft]] #each pft is a first-level list element.
  
  #currently, variable extraction is hard-coded. Have to specify different variables and create objects for them by hand.
  gppall<-pft.dat$GPPAll
  nppall<-pft.dat$NPPAll
  
  vegc.l<-pft.dat$VegCarbon$Leaf
  vegc.w<-pft.dat$VegCarbon$Stem
  vegc.r<-pft.dat$VegCarbon$Root
  vegc.sum<-vegc.l+vegc.w+vegc.r
  
  veg.strn.l<-pft.dat$VegStructuralNitrogen$Leaf
  veg.strn.w<-pft.dat$VegStructuralNitrogen$Stem
  veg.strn.r<-pft.dat$VegStructuralNitrogen$Root
  veg.strn.sum<-veg.strn.l+veg.strn.w+veg.strn.r
  
  veg.lbln<-pft.dat$VegLabileNitrogen
  
  #pft.vals vector should include all the target variables you care about (all objects in preceeding code block)
  pft.vals<-c(gppall,nppall,vegc.l,vegc.w,vegc.r,vegc.sum,veg.strn.l,veg.strn.w,veg.strn.r,veg.strn.sum,veg.lbln)
 
  #if there is only one pft, the output is a vector instead of a matrix.
  if(length(dim(pft.vals.mat))>1){ 
  pft.vals.mat[,j]=pft.vals} else {
  pft.vals.mat=pft.vals}
}

#creating a dataframe with pft-specfic variable names. will have a column for Variable (eg nppall0), and one for Value
pft.vars.rep<-rep(pft.vars,num.pfts)
vars<-c(pft.vars.rep)
pft.for.df<-c(rep(pft.nums,each=length(pft.vars)))
vars.numd=paste(vars,pft.for.df,sep='')

pft.vals.vec<-c(pft.vals.mat) #converting matrix to a single vector. 
vals<-c(pft.vals.vec)

obs.vals<-data.frame(Variable=vars.numd,Value=vals)

print("Writing file:")
print(ofname)
write.table(obs.vals,ofname,row.names=FALSE,quote=FALSE,sep=',')






